ord_schema/frozen_message_test.py

```diff
@@ -0,0 +1,116 @@
+        self.assertTrue(hasattr(frozen, 'rpm'))  # Wrong!
+
+    def test_access_optional_scalar(self):
+        frozen = frozen_message.FrozenMessage(
+            reaction_pb2.Percentage(value=12.3))
+        self.assertTrue(hasattr(frozen, 'value'))
+        self.assertAlmostEqual(frozen.value, 12.3, places=6)
+        self.assertFalse(hasattr(frozen, 'precision'))
+        with self.assertRaises(AttributeError):
+            _ = frozen.precision
+
+    def test_access_submessage(self):
+        frozen = frozen_message.FrozenMessage(reaction_pb2.Reaction())
+        self.assertFalse(hasattr(frozen, 'setup'))
+        with self.assertRaises(AttributeError):
+            _ = frozen.setup
+
+    def test_access_repeated_scalar(self):
+        message = reaction_pb2.ReactionProduct()
+        message.analysis_identity.append('test')
+        frozen = frozen_message.FrozenMessage(message)
+        self.assertLen(frozen.analysis_identity, 1)
+        self.assertEqual(frozen.analysis_identity[0], 'test')
+        with self.assertRaises(IndexError):
+            _ = frozen.analysis_identity[1]
+        self.assertEmpty(frozen.analysis_yield)
+        with self.assertRaises(IndexError):
+            _ = frozen.analysis_yield[0]
+
+    def test_access_repeated_submessage(self):
+        message = reaction_pb2.Reaction()
+        message.observations.add(comment='test')
+        frozen = frozen_message.FrozenMessage(message)
+        self.assertLen(frozen.observations, 1)
+        self.assertEqual(frozen.observations[0].comment, 'test')
+        with self.assertRaises(IndexError):
+            _ = frozen.observations[1]
+        self.assertEmpty(frozen.outcomes)
+        with self.assertRaises(IndexError):
+            _ = frozen.outcomes[0]
+
+    def test_access_map_value(self):
+        message = reaction_pb2.Reaction()
+        message.inputs['test'].addition_order = 1
+        frozen = frozen_message.FrozenMessage(message)
+        self.assertTrue(hasattr(frozen, 'inputs'))
+        self.assertIn('test', frozen.inputs)
+        self.assertTrue(hasattr(frozen.inputs['test'], 'addition_order'))
+        self.assertEqual(frozen.inputs['test'].addition_order, 1)
+        self.assertNotIn('missing', frozen.inputs)
+        with self.assertRaises(KeyError):
+            _ = frozen.inputs['missing']
+
+    def test_set_scalar_value(self):
+        frozen = frozen_message.FrozenMessage(
+            reaction_pb2.StirringConditions.StirringRate())
+        with self.assertRaises(dataclasses.FrozenInstanceError):
+            frozen.rpm = 123
+
+    def test_set_submessage(self):
+        frozen = frozen_message.FrozenMessage(reaction_pb2.Reaction())
+        with self.assertRaises(AttributeError):
+            frozen.setup.automation_platform = 'test'
+        with self.assertRaises(AttributeError):
+            frozen.setup.CopyFrom(
+                reaction_pb2.ReactionSetup(automation_platform='test'))
+
+    def test_set_repeated_scalar(self):
```

Comment:
Function name doesn't match test

Reply:
Changed to `test_add_repeated_scalar`; I was going for consistency with the other names, but this is better.